### PR TITLE
[VEN-1902]: Prime Integration

### PR DESCRIPTION
**Pull request by narayanprusty**
## Description

Resolves #

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -986,6 +1109,18 @@ contract Comptroller is
...
+    /**
+     * @notice Sets the prime token contract for the comptroller
+     * @param _prime Address of the Prime contract
+     */
+    function setPrimeToken(IPrime _prime) external onlyOwner {
+        ensureNonzeroAddress(address(_prime));
+
+        IPrime oldPrime = prime;
+        prime = _prime;
+        emit NewPrimeToken(oldPrime, _prime);
```
Comment (chechu) [inline contracts/Comptroller.sol:1121]: We could emit the event first to save some gas

File: tests/hardhat/Prime.ts [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+  PrimeScenario,
+  ResilientOracleInterface,
+  UpgradeableBeacon,
+  VToken,
+  XVSStoreScenario,
+  XVSVaultScenario,
+} from "../../typechain";
+import { deployVTokenBeacon, makeVToken } from "./util/TokenTestHelpers";
+
+// import { convertToUnit } from "../../../helpers/utils";
```
Comment (chechu) [inline tests/hardhat/Prime.ts:26]:
```suggestion
```

File: tests/hardhat/Prime.ts [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+  await primeScenario.setLimit(1000, 1000);
+  await primeScenario.addMarket(comptrollerProxy.address, vUSDT.address, bigNumber18.mul("1"), bigNumber18.mul("1"));
+  await primeScenario.addMarket(comptrollerProxy.address, vLINK.address, bigNumber18.mul("1"), bigNumber18.mul("1"));
+  await comptrollerProxy.setPrimeToken(primeScenario.address);
+  await primeScenario.togglePause();
+
+  return {
+    oracle: fakePriceOracle,
+    accessControl,
+    // comptrollerLens,
```
Comment (chechu) [inline tests/hardhat/Prime.ts:240]:
```suggestion
```

File: tests/hardhat/Prime.ts
Comment (chechu) [inline tests/hardhat/Prime.ts:1]: I would say we don't need to duplicate the tests already included in the `venus-protocol` package. The focus in this test should be on the integration with the Comptroller, which is the only new feature added. Hopefully, we can remove the dependency with the solidity compiler 0.5.16 too.

File: deployments/sepolia.json
Comment (chechu) [inline deployments/sepolia.json:1]: I don't think we should deploy to sepolia the XVSVault related contracts from this PR

File: contracts/Comptroller.sol [hunk 4]
```diff
@@ -327,6 +345,98 @@ contract Comptroller is
...
 
+    /**
+     * @notice Validates redeem, accrues interest and updates score in prime. Reverts on rejection. May emit logs.
+     * @param vToken Asset being redeemed
+     * @param redeemer The address redeeming the tokens
+     * @param redeemAmount The amount of the underlying asset being redeemed
+     * @param redeemTokens The number of tokens being redeemed
+     */
+    function redeemVerify(address vToken, address redeemer, uint256 redeemAmount, uint256 redeemTokens) external {
+        require(redeemTokens != 0 || redeemAmount == 0, "redeemTokens zero");
```
Comment (Debugger022) [inline contracts/Comptroller.sol:356]: Should we use custom error here instead of require statement.

**Review by chechu — DISMISSED** [review]

**Review by chechu — APPROVED** [review]

**Pull request merged**